source/schema/avro.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/jackc/pgx/v5/pgtype"
+)
+
+const (
+	avroNS = "conduit.postgres"
+)
+
+var Avro = &avroExtractor{
+	pgMap: pgtype.NewMap(),
+	avroMap: map[string]*avro.PrimitiveSchema{
+		"bool":    avro.NewPrimitiveSchema(avro.Boolean, nil),
+		"bytea":   avro.NewPrimitiveSchema(avro.Bytes, nil),
+		"float4":  avro.NewPrimitiveSchema(avro.Float, nil),
+		"float8":  avro.NewPrimitiveSchema(avro.Double, nil),
+		"int8":    avro.NewPrimitiveSchema(avro.Long, nil),
+		"int4":    avro.NewPrimitiveSchema(avro.Int, nil),
+		"int2":    avro.NewPrimitiveSchema(avro.Int, nil),
+		"text":    avro.NewPrimitiveSchema(avro.String, nil),
+		"varchar": avro.NewPrimitiveSchema(avro.String, nil),
+		"timestamptz": avro.NewPrimitiveSchema(
+			avro.Long,
+			avro.NewPrimitiveLogicalSchema(avro.TimestampMicros),
+		),
+		"timestamp": avro.NewPrimitiveSchema(
+			avro.Long,
+			avro.NewPrimitiveLogicalSchema(avro.LocalTimestampMicros),
+		),
+		"date": avro.NewPrimitiveSchema(
+			avro.Int,
+			avro.NewPrimitiveLogicalSchema(avro.Date),
+		),
+		"uuid": avro.NewPrimitiveSchema(
+			avro.String,
+			avro.NewPrimitiveLogicalSchema(avro.UUID),
+		),
+	},
+}
+
+type avroExtractor struct {
+	pgMap   *pgtype.Map
+	avroMap map[string]*avro.PrimitiveSchema
+}
+
+func (a avroExtractor) ExtractLogrepl(rel *pglogrepl.RelationMessage, row *pglogrepl.TupleData) (avro.Schema, error) {
+	var (
+		fields []pgconn.FieldDescription
+		values []any
+	)
+
+	for i, tuple := range row.Columns {
+		fields = append(fields, pgconn.FieldDescription{
+			Name:        rel.Columns[i].Name,
+			DataTypeOID: rel.Columns[i].DataType,
+		})
+
+		v, err := a.decodeColumnValue(rel.Columns[i], tuple.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		values = append(values, v)
+	}
+
+	return a.Extract(rel.RelationName, fields, values)
+}
+
+func (a *avroExtractor) Extract(name string, fields []pgconn.FieldDescription, values []any) (avro.Schema, error) {
+	var avroFields []*avro.Field
+
+	for i, f := range fields {
+		t, ok := a.pgMap.TypeForOID(f.DataTypeOID)
+		if !ok {
+			return nil, fmt.Errorf("field %q with OID %d cannot be resolved", f.Name, f.DataTypeOID)
+		}
+
+		s, err := a.extractType(t, values[i])
+		if err != nil {
+			return nil, err
+		}
+
+		af, err := avro.NewField(f.Name, s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create avro field %q: %w", f.Name, err)
+		}
+
+		avroFields = append(avroFields, af)
+	}
+
+	slices.SortFunc(avroFields, func(a, b *avro.Field) int {
+		return cmp.Compare(a.Name(), b.Name())
+	})
+
+	sch, err := avro.NewRecordSchema(name, avroNS, avroFields)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create avro schema: %w", err)
+	}
+
+	return sch, nil
+}
+
+func (a *avroExtractor) extractType(t *pgtype.Type, val any) (avro.Schema, error) {
+	if ps, ok := a.avroMap[t.Name]; ok {
+		return ps, nil
+	}
+
+	switch tt := val.(type) {
+	case pgtype.Numeric:
+		// N.B.: Default to 38 positions and pick the exponent as the scale.
```

Comment:
Any specific reason for 38?

Reply:
@hariso In this case it is the default scale in Snowflake. In practice because numeric types are limited by the pgx driver to  int64/float64 (at 8 bytes), which really can be accomplished with this being set to 19 (rather than the double 38) 😅 

Now, when a fixed schema is defined with `38` as size, this is the size in bytes. The avro library will use big.Rat as way to encode/decode decimals, so the size will need to be matched to the rational encoded in bytes. It's great.. 38 in this case is absolutely random and it should be ok for now. The way to do this is to set it to the max possible size of big.Rat (which can be excessive) lol

Reply:
Thanks for the explanation! It might be good to just paste this into a comment in the code, as a refresher for future us.:)

Reply:
@hariso yea will do.

Reply:
see 4df7f87